### PR TITLE
Implement get_node with a get_node_raw

### PR DESCRIPTION
**Pull request by Desour**
* Goal: Improve performance of `get_node`.
* How: `get_node` is implemented in builtin with a `get_node_raw(x, y, z) -> content, param1, param2, pos_ok`.
* Why is this beneficial: In master, `get_node` calls lua functions to push nodes, and to read vectors. This is much faster if done in lua.

Flamegraphs:
master:
master
PR:
pr

Rough benchmark results:
master: 450-850 ms
PR: around 270-300 ms (it's *much* more persistent, idk why)

Benchmark function adopted from #14225 by @sfence. (Added you as co-author. :))

## To do

This PR is a Ready for Review.

## How to test

* `/bench_bulk_get_node`
* TODO: Do we have unit tests for `get_node` and `get_node_or_nil`?

Here's some instructions I've noted down at some point to disable cpu scaling, for better testing:
<details>

```
info:
	https://www.kernel.org/doc/html/v5.11/admin-guide/pm/cpufreq.html
	https://wiki.archlinux.org/index.php/CPU_frequency_scaling


check:
	sudo cpupower frequency-info
	watch grep \"cpu MHz\" /proc/cpuinfo


cpu-scaling off:
	sudo sh -c "echo 0 > /sys/devices/system/cpu/cpufreq/boost"
	sudo cpupower frequency-set -g userspace
	sudo cpupower frequency-set -f 1600MHz

cpu-scaling on:
	sudo cpupower frequency-set -g schedutil
	sudo sh -c "echo 1 > /sys/devices/system/cpu/cpufreq/boost"
```

</details>

### CODE REVIEW

File: src/script/lua_api/l_env.cpp [hunk 1]
```diff
@@ -324,39 +324,25 @@ int ModApiEnv::l_swap_node(lua_State *L)
...
+// get_node_raw(x, y, z) -> content, param1, param2, pos_ok
+int ModApiEnv::l_get_node_raw(lua_State *L)
 {
 	GET_ENV_PTR;
 
 	// pos
-	v3s16 pos = read_v3s16(L, 1);
+	double x = lua_tonumber(L, 1);
+	double y = lua_tonumber(L, 2);
+	double z = lua_tonumber(L, 3);
```
Comment (sfan5) [inline src/script/lua_api/l_env.cpp:335]: why not lua_tointeger?

File: src/script/lua_api/l_env.h [hunk 2]
```diff
@@ -74,13 +74,9 @@ class ModApiEnv : public ModApiEnvBase {
...
 
-	// get_node(pos)
-	// pos = {x=num, y=num, z=num}
-	static int l_get_node(lua_State *L);
-
-	// get_node_or_nil(pos)
-	// pos = {x=num, y=num, z=num}
-	static int l_get_node_or_nil(lua_State *L);
+	// get_node_raw(x, y, z) -> content, param1, param2, pos_ok
+	// Used to implement get_node and get_node_or_nil in lua.
```
Comment (sfan5) [inline src/script/lua_api/l_env.h:78]:
```suggestion
	// Used to implement get_node and get_node_or_nil in lua.
	// This is still faster than doing it from C++ even with optimized pushnode
```

File: src/script/lua_api/l_env.cpp (hunk 1 above)
Comment (Desour) [reply src/script/lua_api/l_env.cpp:335]: Exact rounding matters. Mods do things such as `get_node(node_pos:offset(0,-0.5,0))`.

(Would be nice to deprecate this some day.)

Comment (sfan5) [reply src/script/lua_api/l_env.cpp:335]: I see.
suggestion: `// mirrors implementation of read_v3s16 (which does exact rounding)`

**Comment by Desour** [comment]
Added the comments.

File: builtin/common/item_s.lua [hunk 3]
```diff
@@ -239,3 +239,18 @@ if core.set_read_node and core.set_push_node then
 	core.set_push_node(push_node)
 	core.set_push_node = nil
 end
+
+if INIT == "game" then
```
Comment (sfan5) [inline builtin/common/item_s.lua:243]: item.lua is only imported for the game env so that's a better place for this

**Review by sfan5 — APPROVED** [review]
LGTM otherwise

**Review by SmallJoker — COMMENTED** [review]
| Time master / ms | Time PR / ms
-- | -- | --
  | 314.16 | 79.64
  | 290.42 | 81.25
  | 254.11 | 80.21
  | 193.72 | 80.7
  | 184.71 | 80.32
  | 329.17 | 80.05
  | 251.46 | 80.42
  | 301.68 | 79.2
  | 197.04 | 79.23
  | 197.5 | 80.82
  |   |  
average | 251.4 | 80.2
**MEDIAN** | **252.8** | **80.3**

I find it strange that there's this much difference. `push_node` already uses Lua functions to construct the table from arguments (`core.set_push_node`). So apparently by moving the table constructor after "returning" from C++ makes such big difference? Why's that?

**Comment by Desour** [comment]
In master we call lua functions from C++ to read vector and push node values. The C->Lua call overhead is probably quite big. And also we always need to look up the functions to call first, which requires a bunch of Lua C api calls.

**Review by SmallJoker — APPROVED** [review]
Works well. Cannot complain.

**Comment by Desour** [comment]
>Benchmark function adopted from https://github.com/minetest/minetest/pull/14225 by @sfence. (Added you as co-author. :))

Somehow that co-authorship didn't make it into the commit. Did github squash it away?

**Comment by SmallJoker** [comment]
Sorry @Desour. I totally missed that remark and removed it alongside the other commit titles upon merge. GitHub is not to blame.

**Comment by kno10** [comment]
Would it make sense to do the same change in the voxel manipulator API?

**Comment by Desour** [comment]
Would make sense imo, also to make the implementations more consistent.
(Also maybe for `set_node_at()`, which is much simpler and cheaper than `core.set_node`.)

I'd also not not necessarily be against a lower level API. For example a `get_data_at_i(i)`, that works similar to `get_data()[i]`, would make sense imo.

But idk about other coredevs' opinions.

**Pull request merged**